lib/org/chromium/webidl/ParserRunner.js

```diff
@@ -8,45 +8,78 @@ foam.CLASS({
+    'org.chromium.webidl.BaseParser',
+    'org.chromium.webidl.ast.Definition',
+    'org.chromium.webidl.Parser as DefaultParser',
+  ],
 
   documentation: 'Runnable Box that parsers IDLFileContents into ASTs.',
 
-  methods: [
-    function run(message) {
-      if (this.validateMessage(message)) return;
+  properties: [
+    {
+      class: 'Class',
+      documentation: 'Class of the parser used to parse the IDL files.',
+      name: 'Parser',
+      factory: function() {
+        // TODO: Replace with a smarter method of choosing a parser.
+        return this.DefaultParser;
+      },
+    },
+    {
+      class: 'String',
+      name: 'ioRelationshipType',
+      documentation: 'The n:m relationship type of input-to-output.',
+      value: '1:many',
+    },
+    {
+      class: 'Class',
+      documentation: 'Type of input parameter of run().',
+      name: 'inputType',
+      factory: function() {
+        return this.BaseIDLFile;
+      },
+    },
+    {
+      class: 'Class',
+      documentation: 'Type of output values produced by run().',
+      name: 'outputType',
+      factory: function() {
+        return this.Definition;
+      },
+    },
+    {
+      documentation: 'An instance of the parser class.',
+      name: 'parser_',
+      factory: function() {
+        return this.Parser.create();
+      },
+    },
+  ],
 
-      // Verify all required arguments are present.
-      if (!message.idlFile || !message.parserClass)
-        return this.error(this.fmtErrorMsg(this.MISSING_PROPS));
+  methods: [
+    function init() {
+      foam.assert(this.BaseParser.isInstance(this.parser_));
+      this.SUPER();
+    },
+    function run(file) {
+      if (this.validateMessage(file)) return;
 
-      var file = message.idlFile;
-      var parserClass = message.parserClass;
       // Prepare metadata to be injected by parser.
       var metadata = Array.isArray(file.id) ? file.id.join('/') : file.id;
-
-      // Load appropriate parser depending on configuration.
-      // TODO: Replace injected string with a safer method, such as a
-      // helper method that chooses parser based on URL.
-      var parser = parserClass.create();
-      foam.assert(this.BaseParser.isInstance(parser),
-        'The class provided is not an instance of BaseParser.');
-
-      var results = parser.parseString(file.contents, metadata);
+      var results = this.parser_.parseString(file.contents, metadata);
 
       // Determine if parse of entire file was successful.
       if (!Array.isArray(results.value) ||
           file.contents.length !== results.pos) {
-        var metadata = file.metadata || {};
-        var id = `
-          Repository: ${metadata.repository}
-          Path: ${metadata.path}
-          Revision: ${metadata.revision}`;
-
-        this.error(new Error(`Incomplete parse on file. Metadata: ${id}`));
+        this.error(this.fmtErrorMsg(
+            `Incomplete parse on file. Metadata: ${metadata}`));
```

Comment:
Metadata is plural. Can we use some other way of describing this? Is it a URL? Pseudo-URL? ID?